src/dll/Systems/StateSystem.hpp

```diff
@@ -1,45 +1,116 @@
 
-    bool OnEnter(RED4ext::EGameStateType aStateType, RED4ext::CGameApplication* aApp);
-    bool OnUpdate(RED4ext::EGameStateType aStateType, RED4ext::CGameApplication* aApp);
-    bool OnExit(RED4ext::EGameStateType aStateType, RED4ext::CGameApplication* aApp);
+    template<std::derived_from<RED4ext::IGameState> T>
+    bool Wrap(GameStateHook<T>& aHooked, StateSystemAction aAction, T* aThis, RED4ext::CGameApplication* aApp);
 
 private:
+    using Func_t = bool (*)(RED4ext::CGameApplication&);
+
     struct StateItem
     {
         std::shared_ptr<PluginBase> plugin;
         Func_t func;
     };
 
+    struct StateAction
+    {
+        std::wstring_view GetName() const;
+
+        void Shutdown();
+
+        std::wstring_view name;
+        std::vector<StateItem> onBefore;
+        std::vector<StateItem> onAfter;
+    };
+
     struct State
     {
-        std::list<StateItem> onEnter;
-        std::list<StateItem> onUpdate;
-        std::list<StateItem> onExit;
+        std::wstring_view GetName() const;
+
+        void Shutdown();
+
+        std::wstring_view name;
+        StateAction onEnter = {.name = L"Enter"};
+        StateAction onTick = {.name = L"Tick"};
+        StateAction onExit = {.name = L"Exit"};
     };
 
-    State* GetStateByType(RED4ext::EGameStateType aStateType);
+    template<std::derived_from<RED4ext::IGameState> T>
+    bool ExecuteHookedAction(GameStateHook<T>& aHooked, StateSystemAction aAction, T* aThis,
+                             RED4ext::CGameApplication* aApp);
 
-    bool Run(std::wstring_view aAction, std::list<StateItem>& aList, RED4ext::CGameApplication* aApp);
+    State* GetStateByType(RED4ext::EGameStateType aType);
+    std::pair<State*, StateAction*> GetStateAndActionByTypes(RED4ext::EGameStateType aType, StateSystemAction aAction);
 
-    State m_baseInitialization;
-    State m_initialization;
-    State m_running;
-    State m_shutdown;
+    void Run(std::wstring_view aAction, std::vector<StateItem>& aList, RED4ext::CGameApplication& aApp);
+
+    State m_baseInitialization = {.name = L"BaseInitialization"};
+    State m_initialization = {.name = L"Initialization"};
+    State m_running = {.name = L"Running"};
+    State m_shutdown = {.name = L"Shutdown"};
 };
+
+template<std::derived_from<RED4ext::IGameState> T>
+inline bool StateSystem::Wrap(GameStateHook<T>& aHooked, const StateSystemAction aAction, T* aThis,
+                              RED4ext::CGameApplication* aApp)
+{
+    auto [state, action] = GetStateAndActionByTypes(aThis->GetType(), aAction);
+    if (!state || !action)
+        return ExecuteHookedAction(aHooked, aAction, aThis, aApp);
+
+    Run(fmt::format(L"{}::OnBefore{}", state->GetName(), action->GetName()), action->onBefore, *aApp);
+    const auto result = ExecuteHookedAction(aHooked, aAction, aThis, aApp);
+    Run(fmt::format(L"{}::OnAfter{}", state->GetName(), action->GetName()), action->onAfter, *aApp);
```

Comment:
This works differently than before (see https://docs.red4ext.com/mod-developers/custom-game-states#games-life-cycle). A plugin might need to block the game from switching states while it’s waiting for resources, doing extra work, or something similar.

I remember us talking about this on Discord, but I still don’t see why plugins shouldn’t be allowed to do that. Plugins are already low-level, so if avoiding CTDs is the main reason for this change, I don’t think that’s a good enough reason, since they can CTD the game with memory access.

My point with GameStates is: if you need them, then you at least know what you should do with them and when.

Reply:
I’m not against `Pre` hook, but if we’re removing a feature just to add this hook, then yeah, I am.

If we are going to remove this feature, then we can just remove the GameStates completely. I find no reason to have plugins register a game state that cannot block the game until its resources are initialized / loaded, since the `Run` stuff is added.

> I dont think the way it was designed was good for just wanting to have one-off things and it definitely wasnt good enough

Why do you say that?

Imo, the old implementation was okayish. Depending on your state, it was executed after (`BaseInit`, `Init`, `Running` or before (`Shutdown`) the game's state.

Reply:
I've looked at `GameStateHook` and there's something I missed before: it prevents calls to original `OnTick` after it returns 'true`. So it shouldn't cause any issues with plugins holding a state.

Reply:
Thinking about it, we could just leave it as it is for now and create an issue to fix it later, either if someone needs it or if we decide to rework it.

About returning an enum, I fully agree.

Reply:
I agree about extended return value.